Manifold/Constraint.swift

```diff
@@ -121,14 +110,22 @@ public func checkForInconsistencies(partition: [Type]) -> (Error?, Substitution)
 }
 
 public func solve(constraints: ConstraintSet) -> Either<Error, Substitution> {
-	let (equivalences, indexByType) = typeGraph(constraints)
-	let graph = reduce(constraints, equivalences) { graph, constraint in
-		constraint.analysis(
-			ifEquality: { t1, t2 in
-				structural(t1, t2, graph) { graph, t1, t2 in
-					(indexByType[t1] &&& indexByType[t2]).map { graph.union($0, $1) } ?? graph
-				}
-			})
+	func findOrAdd(type: Type, inout equivalences: DisjointSet<Type>, inout indices: [Type: Int]) -> Int {
```

Comment:
This suggests to me that `DisjointSet<T>` should probably include an `IndexedDisjointSet<T>` or similar which constructs the index alongside the set.